File: server/src/archives/metadata/anchira.rs
use super::ArchiveMetadata;
use crate::{archives, utils};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct Metadata {
  #[serde(rename = "Title")]
  pub title: String,
  #[serde(rename = "Description")]
  pub description: Option<String>,
  #[serde(rename = "Source", default)]
  pub source: Option<String>,
  #[serde(rename = "URL", default)]
  pub url: Option<String>,
  #[serde(rename = "Artist", default)]
  pub artists: Option<Vec<String>>,
  #[serde(rename = "Circle", default)]
  pub circles: Option<Vec<String>>,
  #[serde(rename = "Magazine", default)]
  pub magazines: Option<Vec<String>>,
  #[serde(rename = "Event", default)]
  pub events: Option<Vec<String>>,
  #[serde(rename = "Publisher", default)]
  pub publishers: Option<Vec<String>>,
  #[serde(rename = "Parody", default)]
  pub parodies: Option<Vec<String>>,
  #[serde(rename = "Tags", default)]
  pub tags: Option<Vec<String>>,
  #[serde(rename = "Thumbnail", default)]
  pub thumbnail: Option<i16>,
  #[serde(rename = "Released", default)]
  pub released: Option<i64>,
}

pub fn add_metadata(info: Metadata) -> ArchiveMetadata {
  let mut archive = ArchiveMetadata {
    title: info.title,
    description: info.description,
    thumbnail: info.thumbnail,
    released_at: utils::map_timestamp(info.released),
    artists: info.artists,
    circles: info.circles,
    magazines: info.magazines,
    events: info.events,
    publishers: info.publishers,
    parodies: info.parodies,
    tags: info
      .tags
      .map(|tags| tags.into_iter().map(|tag| (tag, "".to_string())).collect()),
    ..Default::default()
  };

  let mut sources: Vec<archives::Source> = vec![];

  if let Some(url) = info.url {
    sources.push(archives::Source {
      name: utils::parse_source_name(&url),
      url: Some(url),
    });
  }

  if let Some(url) = info.source {
    sources.push(archives::Source {
      name: utils::parse_source_name(&url),
      url: Some(url),
    });
  }

  if !sources.is_empty() {
    archive.sources = Some(sources);
  }

  archive
}

File: cli/metadata/anchira.ts
import dayjs from 'dayjs';
import YAML from 'yaml';
import { z } from 'zod';

import { ArchiveMetadata } from '../../shared/metadata';
import { stringOrNumberArray } from './schemas';

const metadataSchema = z.object({
	Title: z.string(),
	Description: z.string().optional(),
	Source: z.string().optional(),
	URL: z.string().optional(),
	Artist: stringOrNumberArray,
	Circle: stringOrNumberArray,
	Magazine: stringOrNumberArray,
	Event: stringOrNumberArray,
	Publisher: stringOrNumberArray,
	Parody: stringOrNumberArray,
	Tags: stringOrNumberArray,
	Thumbnail: z.number().optional(),
	Released: z.number().optional(),
});

export default async (content: string, archive: ArchiveMetadata) => {
	const parsed = YAML.parse(content);
	const { data, error } = metadataSchema.safeParse(parsed);

	if (!data) {
		throw new Error(`Failed to parse Anchira metadata: ${error}`);
	}

	archive = structuredClone(archive);

	archive.title = data.Title;
	archive.title = data.Title;
	archive.description = data.Description;
	archive.thumbnail = data.Thumbnail;
	archive.releasedAt = data.Released ? dayjs.unix(data.Released).toDate() : undefined;

	archive.tags = [];

	for (const tag of data.Artist) {
		archive.tags.push({ namespace: 'artist', name: tag });
	}

	for (const tag of data.Circle) {
		archive.tags.push({ namespace: 'circle', name: tag });
	}

	for (const tag of data.Magazine) {
		archive.tags.push({ namespace: 'magazine', name: tag });
	}

	for (const tag of data.Event) {
		archive.tags.push({ namespace: 'event', name: tag });
	}

	for (const tag of data.Publisher) {
		archive.tags.push({ namespace: 'publisher', name: tag });
	}

	for (const tag of data.Parody) {
		archive.tags.push({ namespace: 'parody', name: tag });
	}

	for (const tag of data.Tags) {
		archive.tags.push({ namespace: 'tag', name: tag });
	}

	archive.sources = [];

	if (data.URL) {
		archive.sources.push({ url: data.URL });
	}

	if (data.Source) {
		archive.sources.push({ url: data.Source });
	}

	return archive;
};
